content/posts/building-images-with-img.md

```diff
@@ -0,0 +1,257 @@
+spec:
+  allowPrivilegeEscalation: true
+  allowedProcMountTypes:
+  - Unmasked
+  - Default
+  fsGroup:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
+  hostPID: false
+  hostIPC: false
+  hostNetwork: false
+  privileged: false
+  runAsUser:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
+  seLinux:
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
+  volumes:
+  - 'emptyDir'
+  - 'secret'
+  - 'downwardAPI'
+  - 'configMap'
+  - 'persistentVolumeClaim'
+  - 'projected'
+```
+
+There are a few details worth calling out in this policy:
+1. We use the `runAsUser` specification to ensure our `img` Pod runs as a non-zero/non-root user.
+2. `privileged` is set to `false`. However, it's worth noting that `allowPrivilegeEscalation` must be set to true for `img` to execute commands properly.
+3. For `img` to run properly, both `seccomp` and `AppArmor` profiles must be set to `unconfined`, which is [required by `runc`](https://github.com/genuinetools/img#running-with-docker).
+4. We include `allowedProcMountTypes` with both `Unmasked` and `Default` as accepted values - more on this in a minute.
+
+Next, we're going to attach our `img` `ServiceAccount` to a corresponding `Role` with a `RoleBinding`. This will allow our `img` `Pod` to use the custom `PodSecurityPolicy` we just created.
+```yaml
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: img
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
+  verbs:     ['use']
+  resourceNames:
+  - img
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["create","delete","get","list","patch","update","watch"]
+
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: img
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: img
+subjects:
+- kind: ServiceAccount
+  name: img
+```
+
+Finally, here is our `Pod` definition:
+```yaml
+# based on https://blog.jessfraz.com/post/building-container-images-securely-on-kubernetes/
+apiVersion: v1
+kind: Pod
+metadata:
+  name: img
+  labels:
+    run: img
+  annotations:
+    container.apparmor.security.beta.kubernetes.io/img: unconfined
+    container.seccomp.security.alpha.kubernetes.io/img: unconfined
+spec:
+  securityContext:
+    runAsUser: 1000
+  serviceAccountName: img
+  containers:
+  - name: img
+    image: gcr.io/melgin/gcloud-img:0.1.1
+    imagePullPolicy: Always
+    command:
+    - cat
+    tty: true
+    # securityContext:
+    #   procMount: Unmasked
+    volumeMounts:
+    - name: docker-config
+      mountPath: /.docker/
+    - name: gcloud-config
+      mountPath: /.config/gcloud
+    - name: cache-volume
+      mountPath: /tmp/
+  volumes:
+  - name: docker-config
+    emptyDir: {}
+  - name: gcloud-config
+    emptyDir: {}
+  - name: cache-volume
+    emptyDir: {}
+  restartPolicy: Never
+  
+```
+
+Again, a few items to focus on here:
```

Comment:
Add a new line here, because in the Preview is not showing the numbered bullets as a list.